[1단계 - 로또 구현] 가콩(최가빈) 미션 제출합니다.

gabean13 opened the pull request:
안녕하세요 리뷰어님 ☺️

7기 BE 가콩(최가빈)입니다.
BE 호떡(민경윤) 크루와 페어 프로그래밍을 진행했습니다!

### 궁금한 점

- getter를 안만들려고 노력했는데 출력할때나, 비지니스 로직을 구현할 때 결국 getter가 필요해서 사용을 하게 된 경우가 많습니다. 한가지 객체가 너무 많은 일을 하는 것 같아서 분리를 하다보니 getter가 계속 필요했습니다.
예를 들어, 일급 컬렉션 A와 B가 있고 두 컬렉션의 정보로 연산을 수행하는 객체 C가 있을 때 C에서 A와 B의 정보를 필요로 할 때 getter를 사용할 수밖에 없었는데, 이런 경우 어떻게 설계를 해야 getter를 적게 사용하고, 책임을 분리할 수 있을지 궁금합니다.
- LottoResult 테스트를 만들고 싶었는데 UserLotto의 생성자에서 금액 문자열을 받아서 자동으로 생성하는 것이여서 LottoResult의 기능을 테스트 하지 못했습니다. 이런 경우에는 더 세세하게 객체를 분리해야하는것일까요?!
- 객체를 생성하기 위해 입력을 받고 이에 대한 유효성 검사를 할 때, 생성하고자 하는 객체 내부에서 검사할 수도 있고 입력 받을 때(InputValidator)에서 검사하는 방법이 있었습니다. 어떤 기준으로 유효성 검사 방법을 선택해야할지 페어와 의견이 갈리는 경우가 있었습니다. 명확한 기준은 없겠지만 어느 정도 가이드라인이 있을까요?
예를 들어 ‘당첨 번호’가 입력 문자열이 쉼표로 구분된 형태인지, 숫자인지, 6개인지 등의 검사를 InputValidator와 WinningLotto 객체 내부 중 어디서 하는게 좋을지 많은 고민이 있었습니다..!

### 프롤로그 학습로그
https://prolog.techcourse.co.kr/studylogs/3827

시간내서 리뷰해주셔서 감사합니다 😊

gabean13 committed:
docs(README) : 요구사항 명세

gabean13 committed:
feat(Lotto) : 로또 생성 기능 구현

gabean13 committed:
test(Lotto) : 로또 생성 기능 테스트

gabean13 committed:
feat(WinningLotto) : 당첨 번호 범위 검사 기능 구현

gabean13 committed:
feat(WinningLotto) : 당첨 번호 중복 검사 기능 구현

gabean13 committed:
feat(WinningLotto) : 보너스 볼 유효성 검사 구현

gabean13 committed:
feat(view) : 구매 금액 입력 및 검증 기능 구현

gabean13 committed:
feat(view) : 사용자 구매 로또 오름차순 출력 기능 구현

gabean13 committed:
feat(input) : 당첨 번호 입력 및 검증 기능 구현

gabean13 committed:
feat(input) : 보너스 볼 입력 및 검증 기능 구현

gabean13 committed:
feat(Rank) : 로또 당첨 등수 계산 기능 구현

gabean13 committed:
feat(LottoResult) : 당첨 통계, 수익률 출력

gabean13 committed:
feat : 오류 안내 메세지 추가

gabean13 committed:
style: 코드 정렬

gabean13 committed:
refactor(Controller) : 패키지 이동 및 불필요한 주석 제거

hyunssooo commented on README.md:
```diff
@@ -1,3 +1,36 @@
-로또 미션 저장소
+### 추가 기능 요구사항
+1. 입력이 잘못될 경우 에러 메시지를 출력하고 그 부분부터 다시 입력받는다
+
```
요구 사항 정리 잘 해주셨습니다! :)
마크다운 문법을 활용해 체크리스트를 작성해 보는 것은 어떨까요? 
체크리스트를 기반으로 기능 구현을 해보시면 좋을 것 같아요. :) commit도 기능 단위로 남길 수 있겠네요. 

- [x] 요구사항 ~~~~
- [ ] 요구사항 ~~~~

```
기능 목록 및 commit 로그 요구사항
- 기능을 구현하기 전에 README.md 파일에 구현할 기능 목록을 정리해 추가한다.
- git의 commit 단위는 앞 단계에서 README.md 파일에 정리한 기능 목록 단위로 추가한다
```

gabean13 replied:
네!! 이렇게 README 수정했습니다. 앞으로 커밋도 리드미에 작성한 기능단위로 남겨볼 수 있도록 해보겠습니다

hyunssooo commented on src/main/java/model/Rank.java:
```diff
@@ -0,0 +1,57 @@
+
+    public static Rank getRank(WinningLotto winningLotto, LottoDto lottoDto) {
+        int duplicateNumber = getDuplicateNumber(winningLotto, lottoDto);
+        for (Rank rank : values()) {
+            if (rank.matchNumber == 5 && isBonusMatch(winningLotto.getBonus(), lottoDto)) {
+                return SECOND;
+            }
+            if (rank.matchNumber == duplicateNumber) {
+                return rank;
+            }
+        }
```
https://techcourse.woowahan.com/s/W66QswFx/ls/mLzgp1Cg
프로그래밍 요구 사항을 지켜볼 수 있을까요?

규칙 1: 한 메서드에 오직 한 단계의 들여쓰기(indent)만 한다.

gabean13 replied:
넵 그래서 이렇게 수정을 해봤습니다.
```    
public static Rank getRank(WinningLotto winningLotto, LottoDto lottoDto) {
        int duplicateNumber = getDuplicateNumber(winningLotto, lottoDto);
        if (duplicateNumber == BONUS_REQUIRED_RANK_NUMBER && isBonusMatch(winningLotto.getBonus(), lottoDto)) {
            return SECOND;
        }

        return Arrays.stream(values())
                .filter(rank -> rank.matchNumber == duplicateNumber)
                .findFirst()
                .orElse(FAIL);
    }
```
원래는 if를 판단하는 애들별로 메서드를 분리하려고 했는데 그렇게 하니 if 메서드에서 return 값이 없어서 null인 경우에는 continue를 하게 해야해서 오히려 가독성이 떨어져서 if와 stream을 이용하는 방법을 이용했는데 이 경우도 가독성이 기존보다 떨어지는 것 같다고 느꼈습니다.
혹시 더 좋은 방안이 있을까요? 😢

hyunssooo replied:
음 어느 부분에서 가독성이 떨어진다고 느끼셨나요?

gabean13 replied:
해당 코드가 가독성이 떨어진다기 보다, 전에 코드가 더 가독성이 낫다고 느꼈습니다.
전에는 for문안에서 돌면서 원하는 rank가 있으면 반환하는 것이고, 현재 수정된 코드는 second를 처리하는 부분과 나머지 부분이 나눠져있어서 그렇게 느낀 것 같습니다.

hyunssooo commented on src/main/java/model/Rank.java:
```diff
@@ -0,0 +1,57 @@
+    public static Rank getRank(WinningLotto winningLotto, LottoDto lottoDto) {
+        int duplicateNumber = getDuplicateNumber(winningLotto, lottoDto);
+        for (Rank rank : values()) {
+            if (rank.matchNumber == 5 && isBonusMatch(winningLotto.getBonus(), lottoDto)) {
+                return SECOND;
+            }
```
> getter를 안만들려고 노력했는데 출력할때나, 비지니스 로직을 구현할 때 결국 getter가 필요해서 사용을 하게 된 경우가 많습니다. 한가지 객체가 너무 많은 일을 하는 것 같아서 분리를 하다보니 getter가 계속 필요했습니다.
예를 들어, 일급 컬렉션 A와 B가 있고 두 컬렉션의 정보로 연산을 수행하는 객체 C가 있을 때 C에서 A와 B의 정보를 필요로 할 때 getter를 사용할 수밖에 없었는데, 이런 경우 어떻게 설계를 해야 getter를 적게 사용하고, 책임을 분리할 수 있을지 궁금합니다.

getter를 극단적으로 사용하지 않는 연습을 하는 것은 학습 측면에서 괜찮다고 생각합니다. 하지만 필요한 경우에는 비즈니스 로직에서도 사용할 수 있다고 생각해요. 🙂

말씀하신 부분에서는 C 객체에 A, B 객체를 넘기고 C 객체가 가진 정보를 다시 A, B 객체에게 넘겨 위임할 수 있을 것 같아요. 

로또를 예를 들면 설계에 따라서 다를 수 있겠지만 Lotto 객체에게 숫자 하나를 넘겨 일치하는 숫자가 있는지 물어 볼 수 있을 것 같아요. 혹은 6개의 숫자를 넘겨 일치하는 숫자가 몇개 있는지 Lotto 객체에게 물어볼 수 있을 것 같습니다. :) 숫자 6개를 넘기고 보너스 숫자를 넘긴다면 자연스럽게 결과값으로 Rank 객체를 생성할 수 있지 않을까요?

gabean13 replied:
네 답변 감사합니다 항상 도메인의 역할이 어디까지인지 고민해보겠습니다!!

gabean13 replied:
말씀 주신대로 각 로또마다의 당첨 결과를 계산하는 로직은 Rank보다 Lotto의 책임이 맞다는 생각과, 그렇게 한다면 UserLotto의 getter가 필요없어진다는 점에서 Lotto객체에 WinningLotto를 받아서 결과를 계산한뒤 Rank를 반환하도록 수정했습니다!!

hyunssooo commented on src/main/java/model/LottoResult.java:
```diff
@@ -0,0 +1,51 @@
+import java.util.EnumMap;
+import java.util.Map.Entry;
+
+public class LottoResult {
+    private final EnumMap<Rank, Integer> ranks = new EnumMap<>(Rank.class);
+    private final double profitRate;
+
+    public LottoResult(UserLotto userLotto, WinningLotto winningLotto) {
+        initRank();
+        calculateRanks(userLotto, winningLotto);
+        profitRate = calculateProfitRate(userLotto.getPurchaseAmount(), getProfit());
+    }
```
> LottoResult 테스트를 만들고 싶었는데 UserLotto의 생성자에서 금액 문자열을 받아서 자동으로 생성하는 것이여서 LottoResult의 기능을 테스트 하지 못했습니다. 이런 경우에는 더 세세하게 객체를 분리해야하는것일까요?!

LottoResult 객체 생성자에서 UserLotto 객체를 받지 않으면 되지 않을까요?

gabean13 replied:
결과를 계산한 뒤 수익률까지 계산하는 순서를 지키고 싶은 것과 final로 선언하여 생성자에서 값들을 할당하게 해주기 위해 저렇게 설계하게 된것같습니다.

profitRate를 final로 선언하지 않고, 외부에서 calculateLottoResult를 호출해서 calculateRanks와 calcaulateProfitRate를 계산하게 하는 방식이 더 좋은 방법일까요??

hyunssooo replied:
개인적으로는 좋은 방법은 아닌 것 같아요. 필드에 선언된 ranks 하나면 수익률을 구할 수 있지 않나요?

gabean13 replied:
헉 넵 그렇네요 그렇게 수정하겠습니다

hyunssooo commented on src/main/java/view/InputValidator.java:
```diff
@@ -0,0 +1,34 @@
+package view;
+
+public class InputValidator {
+    private static final String PURCHASE_AMOUNT_INPUT_ERROR = "1000원 단위의 금액을 입력해주세요.\n";
+    private static final String WINNING_NUMBER_INPUT_ERROR = ",로 구분한 6개의 숫자를 입력해주세요.\n";
+
```
> 객체를 생성하기 위해 입력을 받고 이에 대한 유효성 검사를 할 때, 생성하고자 하는 객체 내부에서 검사할 수도 있고 입력 받을 때(InputValidator)에서 검사하는 방법이 있었습니다. 어떤 기준으로 유효성 검사 방법을 선택해야할지 페어와 의견이 갈리는 경우가 있었습니다. 명확한 기준은 없겠지만 어느 정도 가이드라인이 있을까요?
예를 들어 ‘당첨 번호’가 입력 문자열이 쉼표로 구분된 형태인지, 숫자인지, 6개인지 등의 검사를 InputValidator와 WinningLotto 객체 내부 중 어디서 하는게 좋을지 많은 고민이 있었습니다..!

저는 보통 view에서는 입력 포맷만 확인하고 요구 사항에 대한 검증은 도메인 모델에서 이뤄져야 한다고 생각해요. :)

gabean13 replied:
네! 답변 감사합니다 입력 포맷까지만 확인하는게 InputValidator의 역할인 것 같습니다!
 
purchase amount의 입력 포맷은 숫자여서 해당 부분까지는 InputValidator에서 확인하고 1000원 단위인지는 UserLotto에서 검증하도록 변경하였습니다

hyunssooo commented on src/main/java/model/WinningLotto.java:
```diff
@@ -0,0 +1,89 @@
+public class WinningLotto {
+
+    public static final String NUMBER_REGEX = "^[0-9]*[0-9]$";
+    public static final String WINNING_NUMBERS_INPUT_ERROR_MESSAGE = "당첨 번호는 중복되지 않는 1 이상 45 이하의 정수여야합니다.\n";
+    public static final String BONUS_INPUT_ERROR_MESSAGE = "보너스볼은 당첨 번호와 중복되지 않는 1 이상 45 이하의 정수여야합니다.\n";
```
이 부분 때문에 실패하는 테스트가 있는 것 같아요. :)

gabean13 replied:
헉 감사합니다 해당 부분은 hasMessageContaining 메서드로 변경하였습니다.

궁금한 점이 있습니다. 테스트를 할 때 에러 메세지도 검사하는 것을 권장하는지 궁금합니다.
그리고 검사하는 것을 권장한다면 이렇게 문자열로 넣어 놓는 것이 좋은지, 아니면 상수로 넣어 놓는 것이 좋은지도 궁금합니다.
제가 생각했을 때는 한 클래스에서만 사용하는 상수이면 private으로 선언하는것이 맞다고 생각하기도 하는 동시에, 검사를 할 때 에러 메세지를 하드코딩해서 넣어놓는 방식도 좋지 않은 방식이라고 생각이 되어서 고민이 되었습니다.

hyunssooo replied:
저라면 굳이 메세지까지는 테스트에서 확인하지 않을 것 같아요.

gabean13 replied:
넵 답변 감사합니다!

hyunssooo reviewed (changes requested):
안녕하세요. 가콩!
이번 미션 함께 진행할 지노입니다. :) 
몇 가지 코멘트 남겼는데 확인 부탁드려요!

gabean13 committed:
docs(README) : 기능 요구사항 체크박스로 변경

gabean13 committed:
refactor(Rank) : 한 단계 들여쓰기만 하도록 코드 수정

gabean13 committed:
refactor(Input) : 구매 금액 검증 방법 분리

- 숫자인지 그 외의 입력인지는 InputValidator에서 확인하도록 변경
- 1000원 단위인지는 UserLotto에서 확인하도록 변경

gabean13 committed:
refactor(UserLotto) : 에러 메세지 상수화

gabean13 committed:
fix(WinningLottoTest) : hasMessage 메서드 hasMessageContaining으로 수정

gabean13 committed:
refactor(Lotto) : lotto rank 결과 계산을 rank에서 lotto 객체에서 하는 것으로 변경

- lotto에서 WinningLotto를 받아 rank를 계산하는 역할까지 부여
- lotto 생성할 때 랜덤 생성 방식 Math.random()에서 Random class이용하는 것으로 변경
- Rank Test 삭제, Lotto Test 추가 테스트 구현

gabean13 committed:
fix(test) : 로직 변경에 따른 테스트 변경

gabean13 committed:
docs(README) : 기능 요구사항 변경에 따른 readme 수정

gabean13 committed:
Merge remote-tracking branch 'origin/step1' into step1

gabean13 commented:
제가 브랜치를 실수해서 에러 메세지 상수화 부분, hasMessage 메서드 변경 커밋 부분이 분기되어서 없애버렸습니다..ㅠㅠ `lotto rank 결과 계산을 rank에서 lotto 객체에서 하는 것으로 변경` 부분부터 수정해서 올려놓았습니다 죄송합니다

hyunssooo commented on src/main/java/dto/LottoDto.java:
```diff
@@ -0,0 +1,31 @@
+import java.util.Collections;
+import java.util.List;
+import model.Lotto;
+
+public record LottoDto(
+        List<Integer> lotto
+) {
```
Dto는 어떤 경우에 사용하시나요? 의존성을 보니 도메인 모델이 의존하는 경우도 있고 도메인 모델의 dto가 의존하고도 있어서요.

hyunssooo commented on src/main/java/dto/LottoDto.java:
```diff
@@ -0,0 +1,31 @@
+        return new LottoDto(lottoNumbers);
+    }
+
+    @Override
+    public String toString() {
+        List<String> lottoStrings = new ArrayList<>(lotto.size());
+        for (int lottoNumber : lotto) {
+            lottoStrings.add(String.valueOf(lottoNumber));
+        }
+        String lottoString = String.join(", ", lottoStrings);
+        return "[" + lottoString + "]";
+    }
```
toString은 보통 디버깅용으로 사용합니다.
출력을 위해서는 getter를 사용하시면 될 것 같아요. 현재는 콘솔 환경이지만 웹 환경이라면 toString은 유효할까요?

hyunssooo commented on src/main/java/model/Lotto.java:
```diff
@@ -0,0 +1,59 @@
+public class Lotto {
+    private static final int BONUS_REQUIRED_RANK_NUMBER = 5;
+    private static final int LOTTO_SIZE = 6;
+
+    private final List<Integer> lotto;
+
```
중복을 허용하지 않는 자료구조는 없을까요?

hyunssooo replied:
1 ~ 45 를 나타내는 클래스도 작성해 볼 수 있지 않을까요?

hyunssooo commented on src/main/java/model/Lotto.java:
```diff
@@ -0,0 +1,59 @@
+
+    private final List<Integer> lotto;
+
+    public Lotto(long currentTime) {
+        this.lotto = generateLotto(currentTime);
+    }
```
로또 생성이 랜덤에 의존하고 있는데요. 랜덤은 테스트하기 어렵습니다. 어떻게 하면 테스트하기 쉽게 변경할 수 있을까요?

hyunssooo commented on src/main/java/model/Lotto.java:
```diff
@@ -0,0 +1,59 @@
+                .orElse(Rank.FAIL);
+    }
+
+    private int getDuplicateNumber(WinningLotto winningLotto) {
+        Set<Integer> union = new HashSet<>();
+        union.addAll(winningLotto.getWinningNumbers());
+        union.addAll(lotto);
+        return 12 - union.size();
+    }
```
12는 어떤 의미일까요?

hyunssooo commented on src/main/java/model/LottoResult.java:
```diff
@@ -0,0 +1,50 @@
+        profitRate = calculateProfitRate(userLotto.getPurchaseAmount(), getProfit());
+    }
+
+    public EnumMap<Rank, Integer> getRanks() {
+        return ranks;
+    }
```
특별한 이유가 아니면 interface를 제공하는 것이 좋습니다 :) 
구현체로 enumMap을 사용하고 외부에서는 map에 의존하게 하는 것은 어떨까요?

hyunssooo commented on src/main/java/model/UserLotto.java:
```diff
@@ -0,0 +1,47 @@
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+public class UserLotto {
+    private static final String PURCHASE_AMOUNT_1000_UNIT_EXCEPTION = "1000원 단위의 금액을 입력해주세요.\n";
+    private final int purchaseNumber;
+    private final List<Lotto> lottos = new ArrayList<>();
```
해당 클래스의 의도는 무엇일까요? purchaseNumber를 필드에 가지고 있어야 하나요?

hyunssooo commented on src/main/java/model/WinningLotto.java:
```diff
@@ -0,0 +1,89 @@
+    public static final String NUMBER_REGEX = "^[0-9]*[0-9]$";
+    public static final String WINNING_NUMBERS_INPUT_ERROR_MESSAGE = "당첨 번호는 중복되지 않는 1 이상 45 이하의 정수여야합니다.\n";
+    public static final String BONUS_INPUT_ERROR_MESSAGE = "보너스볼은 당첨 번호와 중복되지 않는 1 이상 45 이하의 정수여야합니다.\n";
+
+    private final List<Integer> winningNumbers = new ArrayList<>();
```
WinningLotto의 당첨 번호 6개는 Lotto 라고 볼 수 있지 않을까요?

hyunssooo commented on src/main/java/model/WinningLotto.java:
```diff
@@ -0,0 +1,89 @@
+    public static final String BONUS_INPUT_ERROR_MESSAGE = "보너스볼은 당첨 번호와 중복되지 않는 1 이상 45 이하의 정수여야합니다.\n";
+
+    private final List<Integer> winningNumbers = new ArrayList<>();
+    private int bonus;
+
```
WinningLotto의 생성과 별개로 bonus를 채워주는 이유가 따로 있을까요? 생성과 동시에 모든 필드에 할당 될 수는 없을까요?

hyunssooo commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,33 @@
+
+    private final InputValidator inputValidator = new InputValidator();
+
+    public String getPurchaseAmountInput() {
+        String input = getInput();
+        inputValidator.validatePurchaseAmount(input);
+
+        return input;
+    }
```
개인적으로 inputView에서 데이터 타입 정도는 변환해도 괜찮다고 생각해요. 
원시값이나 List 정도를 리턴할 수 있을 것 같아요. 요 부분은 참고만 부탁드려요~

hyunssooo commented on src/test/java/model/LottoTest.java:
```diff
@@ -0,0 +1,55 @@
+class LottoTest {
+    Lotto lotto;
+
+    @BeforeEach
+    void setUp() {
+        //[3, 5, 16, 36, 39, 44]
+        lotto = new Lotto(0);
+    }
```
숫자들이 고정되어 있는 것 같은데요. 유연하게 테스트 가능한 방법을 고민해 보시면 좋을 것 같아요.

hyunssooo commented on src/test/java/model/WinningLottoTest.java:
```diff
@@ -0,0 +1,64 @@
+                .doesNotThrowAnyException();
+    }
+
+    @DisplayName("당첨 번호가 1 이상 45 이하의 정수가 아니면 예외가 발생한다")
+    @ParameterizedTest
+    @ValueSource(strings = {"1, 100, 2, 3, 4, 5", "abc, 1, 2, 3, 4, 5", "-1, 1, 2, 3, 4, 5", "1, 2, 3, 4, 5, 46"})
+    void 당첨_번호가_1_이상_45_이하의_정수가_아니면_예외가_발생한다(String input) {
+        org.assertj.core.api.Assertions.assertThatThrownBy(() -> {new WinningLotto(input);})
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessageContaining("당첨 번호는 중복되지 않는 1 이상 45 이하의 정수여야합니다.");
+    }
```
ParameterizedTest, ValueSource 👍

hyunssooo commented on src/test/java/model/WinningLottoTest.java:
```diff
@@ -0,0 +1,64 @@
+                .doesNotThrowAnyException();
+    }
+
+    @DisplayName("보너스 볼이 1 이상 45 이하의 정수가 아니라면 예외가 발생한다")
+    @ParameterizedTest
+    @ValueSource(strings = {"0", "-1", "46", "100"})
+    void 보너스_볼이_1_이상_45_이하의_정수가_아니라면_예외가_발생한다(String bonus) {
+        WinningLotto winningLotto = new WinningLotto("1, 2, 3, 4, 15, 45");
+        org.assertj.core.api.Assertions.assertThatThrownBy(() -> {winningLotto.setBonus(bonus);})
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessage(BONUS_INPUT_ERROR_MESSAGE);
+    }
```
경계값 테스트 좋습니다 :)

hyunssooo reviewed (approved):
안녕하세요. 가콩! :) 
몇 가지 코멘트 남겼습니다. 다음 step에서 확인 부탁드려요!
이번 step은 미션 사이클을 경험해 보는 것이 중요해 이만 merge 하도록 할게요!
